### ISSUE TITLE
Add extension methods for easily mapping area routes

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#3636]
### The problem of Areas

We're planning on removing *magic link generation* from MVC. This is the feature that causes MVC to only generate a link when we can guarantee that said URL will match an action. We're removing it because it's complicated and doesn't seem to add too much value - users are surprised by the behavior, and we don't want to make it throw.

However, we need to solve the problem that *magic link generation* solves. Namely generating links to areas....

#### Example 1:
```
app.UseMvc(routes =>
{
    routes.MapRoute("Admin/{controller}/{action}/{id?}", defaults: new { area = "Admin" });
    routes.MapRoute("{controller}/{action}/{id?}");
}

@Url.Action("Index", "Home"); // Generates /Admin/Home/Index
@Url.Action("Index", "Home", new { area = "Admin" }); // Generates /Admin/Home/Index
```

#### Example 2:
```
app.UseMvc(routes =>
{
    routes.MapRoute("{controller}/{action}/{id?}");
    routes.MapRoute("Admin/{controller}/{action}/{id?}", defaults: new { area = "Admin" });
}

@Url.Action("Index", "Home"); // Generates /Home/Index
@Url.Action("Index", "Home", new { area = "Admin" }); // Generates /Home/Index?area=Admin
```

In either of these cases, this is not desirable.. You can't use ordering to resolve the problem, because there's no ordering of routes that puts the 'most specific' first. MVC5 solved this problem by **manually filtering the route collection**. This is not a solution that we're willing to pursue in MVC6.

### Something that works

```
app.UseMvc(routes =>
{
    routes.MapRoute(
        "Admin/{controller}/{action}/{id?}",
        defaults: new { area = "Admin" },
        constraints: new { area = "Admin" });

    routes.MapRoute("{controller}/{action}/{id?}");
}

@Url.Action("Index", "Home"); // Generates /Home/Index
@Url.Action("Index", "Home", new { area = "Admin" }); // Generates /Admin/Home/Index
```

**Note that the area route needs to come first in this case** as it's more specific.

### Additionally

We want to provide something simpler that does this for you in the common case:

```
app.UseMvc(routes =>
{
    routes.MapAreaRoute("Admin", "Admin/{controller}/{action}/{id?}"});
    routes.MapRoute("{controller}/{action}/{id?}");
}

@Url.Action("Index", "Home"); // Generates /Home/Index
@Url.Action("Index", "Home", new { area = "Admin" }); // Generates /Admin/Home/Index
```

**Comment by rynowak** [comment]
Self-assigning as I'm working on this already

**Comment by ToddThomson** [comment]
@rynowak As a lone wolf website developer, I construct sites that use area routing techniques that map modules ( class libraries containing controllers and an area subdirectory for all module views ) to area routes. Your solution above is exactly what I've always used ( MapModuleRoute() ).
Thank-you for the razor precompilation support in MVC6 - it now allows me to drop in modules to my applications as needed without having to copy over the file based views.

**Comment by rynowak** [comment]
Thanks :+1: If you feel like there's more stuff you've had to build in this area that should be in the framework, please open specific issues.

6875ee55d3c8632a92432b4b80b139f293f14804

**Comment by ToddThomson** [comment]
@rynowak I was working on attribute routing with areas last Friday. I thought that there might be a nice enhancement for removing the need for manual routes if [Area("SomeArea"] was used with [Route("[area]/[controller]"). If attribute routing could create a route with a default area name "SomeArea" then there would be no need for the manual route. Just an idea as it would be nice to completely remove the need for manual routes.

**Comment by rynowak** [comment]
What you described there sounds like it would work. Can you maybe provide an example of the code you have to write today vs what code you'd like to write?

**Comment by rynowak** [comment]
Fixed by: 6875ee55d3c8632a92432b4b80b139f293f14804
